### ISSUE TITLE
isIncomplete not working as intended

### ISSUE THREAD

**Issue by rwols** [scalameta/metals#1670]
**Describe the bug**

A misreading of the LSP spec perhaps.
When completions are incomplete, the editor will re-request new completions once more characters have been typed by the user. metals should bring back a list with more precise completions. Instead, metals returns nothing. Below are the relevant log lines from SublimeText:

```
:: --> metals-sublime textDocument/completion(57): {'textDocument': {'uri': 'file:///home/raoul/dev/metals/metals/src/main/scala/scala/meta/internal/metals/Cancelable.scala'}, 'position': {'character': 35, 'line': 25}}
:: <<< metals-sublime 57: {'isIncomplete': True, 'items': [{'preselect': True, 'insertTextFormat': 2, 'sortText': '00000', 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Traversable'}, 'label': 'Traversable scala.collection', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/package.Traversable.'}, 'kind': 12, 'filterText': 'Traversable'}, {'kind': 7, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Testing'}, 'label': 'Testing', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/meta/internal/metals/Testing#'}, 'sortText': '00001', 'filterText': 'Testing'}, {'kind': 7, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'TextEdits'}, 'label': 'TextEdits', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/meta/internal/metals/TextEdits#'}, 'sortText': '00002', 'filterText': 'TextEdits'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Thread'}, 'label': 'Thread', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/Thread#'}, 'sortText': '00003', 'filterText': 'Thread'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'ThreadDeath'}, 'label': 'ThreadDeath', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/ThreadDeath#'}, 'sortText': '00004', 'filterText': 'ThreadDeath'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'ThreadGroup'}, 'label': 'ThreadGroup', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/ThreadGroup#'}, 'sortText': '00005', 'filterText': 'ThreadGroup'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'ThreadLocal'}, 'label': 'ThreadLocal', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/ThreadLocal#'}, 'sortText': '00006', 'filterText': 'ThreadLocal'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Throwable'}, 'label': 'Throwable', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/Throwable#'}, 'sortText': '00007', 'filterText': 'Throwable'}, {'kind': 8, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Throwable'}, 'label': 'Throwable', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/package.Throwable#'}, 'sortText': '00008', 'filterText': 'Throwable'}, {'kind': 8, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Traversable'}, 'label': 'Traversable', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/package.Traversable#'}, 'sortText': '00009', 'filterText': 'Traversable'}, {'kind': 8, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'TraversableOnce'}, 'label': 'TraversableOnce', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/package.TraversableOnce#'}, 'sortText': '00010', 'filterText': 'TraversableOnce'}, {'kind': 7, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'TrigramSubstrings'}, 'label': 'TrigramSubstrings', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/meta/internal/metals/TrigramSubstrings#'}, 'sortText': '00011', 'filterText': 'TrigramSubstrings'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'TypeNotPresentException'}, 'label': 'TypeNotPresentException', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/TypeNotPresentException#'}, 'sortText': '00012', 'filterText': 'TypeNotPresentException'}, {'kind': 7, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'EmptyCancelToken'}, 'label': 'EmptyCancelToken', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/meta/internal/metals/EmptyCancelToken#'}, 'sortText': '00013', 'filterText': 'EmptyCancelToken'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'IllegalThreadStateException'}, 'label': 'IllegalThreadStateException', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/IllegalThreadStateException#'}, 'sortText': '00014', 'filterText': 'IllegalThreadStateException'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'InheritableThreadLocal'}, 'label': 'InheritableThreadLocal', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/InheritableThreadLocal#'}, 'sortText': '00015', 'filterText': 'InheritableThreadLocal'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'StackTraceElement'}, 'label': 'StackTraceElement', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'java/lang/StackTraceElement#'}, 'sortText': '00016', 'filterText': 'StackTraceElement'}, {'kind': 9, 'insertTextFormat': 2, 'sortText': '00017', 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Triple'}, 'label': 'Triple scala.Predef', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Predef.Triple.'}, 'deprecated': True, 'filterText': 'Triple'}, {'kind': 7, 'insertTextFormat': 2, 'sortText': '00018', 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Triple'}, 'label': 'Triple', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Predef.Triple#'}, 'deprecated': True, 'filterText': 'Triple'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple1'}, 'label': 'Tuple1 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple1.'}, 'sortText': '00019', 'filterText': 'Tuple1'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple2'}, 'label': 'Tuple2 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple2.'}, 'sortText': '00020', 'filterText': 'Tuple2'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple3'}, 'label': 'Tuple3 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple3.'}, 'sortText': '00021', 'filterText': 'Tuple3'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple4'}, 'label': 'Tuple4 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple4.'}, 'sortText': '00022', 'filterText': 'Tuple4'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple5'}, 'label': 'Tuple5 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple5.'}, 'sortText': '00023', 'filterText': 'Tuple5'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple6'}, 'label': 'Tuple6 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple6.'}, 'sortText': '00024', 'filterText': 'Tuple6'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple7'}, 'label': 'Tuple7 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple7.'}, 'sortText': '00025', 'filterText': 'Tuple7'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple8'}, 'label': 'Tuple8 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple8.'}, 'sortText': '00026', 'filterText': 'Tuple8'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple9'}, 'label': 'Tuple9 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple9.'}, 'sortText': '00027', 'filterText': 'Tuple9'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple10'}, 'label': 'Tuple10 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple10.'}, 'sortText': '00028', 'filterText': 'Tuple10'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple11'}, 'label': 'Tuple11 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple11.'}, 'sortText': '00029', 'filterText': 'Tuple11'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple12'}, 'label': 'Tuple12 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple12.'}, 'sortText': '00030', 'filterText': 'Tuple12'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple13'}, 'label': 'Tuple13 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple13.'}, 'sortText': '00031', 'filterText': 'Tuple13'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple14'}, 'label': 'Tuple14 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple14.'}, 'sortText': '00032', 'filterText': 'Tuple14'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple15'}, 'label': 'Tuple15 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple15.'}, 'sortText': '00033', 'filterText': 'Tuple15'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple16'}, 'label': 'Tuple16 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple16.'}, 'sortText': '00034', 'filterText': 'Tuple16'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple17'}, 'label': 'Tuple17 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple17.'}, 'sortText': '00035', 'filterText': 'Tuple17'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple18'}, 'label': 'Tuple18 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple18.'}, 'sortText': '00036', 'filterText': 'Tuple18'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple19'}, 'label': 'Tuple19 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple19.'}, 'sortText': '00037', 'filterText': 'Tuple19'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple20'}, 'label': 'Tuple20 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple20.'}, 'sortText': '00038', 'filterText': 'Tuple20'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple21'}, 'label': 'Tuple21 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple21.'}, 'sortText': '00039', 'filterText': 'Tuple21'}, {'kind': 9, 'insertTextFormat': 2, 'textEdit': {'range': {'end': {'character': 35, 'line': 25}, 'start': {'character': 34, 'line': 25}}, 'newText': 'Tuple22'}, 'label': 'Tuple22 scala', 'data': {'target': 'file:/home/raoul/Documents/Programming/metals/metals/?id=metals', 'symbol': 'scala/Tuple22.'}, 'sortText': '00040', 'filterText': 'Tuple22'}]}
:: --> metals-sublime textDocument/completion(58): {'textDocument': {'uri': 'file:///home/raoul/dev/metals/metals/src/main/scala/scala/meta/internal/metals/Cancelable.scala'}, 'position': {'character': 36, 'line': 25}}
:: <<< metals-sublime 58: {'isIncomplete': False, 'items': []}
```

**To Reproduce**

Open the file metals/src/main/scala/scala/meta/internal/metals/Cancelable.scala
Put the cursor on line 26.
Type

    var errors = ListBuffer.empty[T

See `Throwable` among the possible completions.
Type

    var errors = ListBuffer.empty[Th

All completions are gone.

**Expected behavior**

The completion list should contain `Throwable`.

**Screenshots**

![Peek 2020-04-28 11-02](https://user-images.githubusercontent.com/2431823/80468769-f257b280-893f-11ea-8f80-8318875c2b30.gif)

**Installation:**
- Editor: Sublime
- Metals version: v0.7.6

**Comment by rwols** [comment]
It's possible that even after knowing that `Th` is the prefix, metals may still set `isIncomplete` to true, which is okay.

**Comment by ayoub-benali** [comment]
I can' reproduce it with version `0.8.4`


![Peek 2020-04-28 11-14](https://user-images.githubusercontent.com/1632384/80469878-7b231e00-8941-11ea-927b-7ee938401f54.gif)

**Comment by rwols** [comment]
I did a `git pull` in my metals-sublime git repo. I think it downloaded 0.8.4, but I still get no completions when isIncomplete is true. How do I verify whether I'm running metals 0.8.4?

Anyway, since it's fixed in 0.8.4 I'll close this. This is more of a setup issue :)

**Comment by olafurpg** [comment]
Thank you for reporting! I'm unable to reproduce. I tried to implement `isIncomplete` according to the same understanding of the spec, that the client should continue to send completion requests as new characters are typed.

Metals completion results include symbols that get auto-imported from the classpath. A query like `Th` may have thousands of hits from the classpath so Metals limits the results to some small number N of results and returns `isIncomplete: true`.

**Comment by olafurpg** [comment]
I don't think anything has changed in this particular functionality recently. However, I get no completion results when the closing square bracket is missing
```
val errors = ListBuffer.empty[Th@@  // Empty completions
val errors = ListBuffer.empty[Th@@] // Non-empty completions
```

**Comment by rwols** [comment]
My gif shows that I have the opening and closing bracket. Whether the closing bracket is present or not shouldn't matter, as it's not part of the prefix, no?

**Comment by ayoub-benali** [comment]
> How do I verify whether I'm running metals 0.8.4?

In the log panel you should see something like this
```
metals-sublime: 2020.04.28 11:38:04 INFO  started: Metals version 0.8.4+57-8dc77885-SNAPSHOT in workspace
```

**Comment by rwols** [comment]
I am indeed running 0.8.4 it seems:

```
metals-sublime: INFO  started: Metals version 0.8.4 in workspace '/home/raoul/Documents/Programming/metals'
```

But the behavior is still present here.

**Comment by ayoub-benali** [comment]
@rwols are you on particular LSP branch ?

**Comment by rwols** [comment]
Actually your gif is not doing exactly what I'm doing @ayoub-benali. You seem to be manually invoking auto-complete.

First type the letter `T`. See completions appear. Then type the letter `h`. Completions disappear.

In fact it doesn't really matter whether I manually invoke auto-complete or not. If I have already typed the letter `T`, then manually invoke auto-complete, and then type the letter `h`, completions also disappear. Perhaps this is fixed in master but not yet released?

**Comment by rwols** [comment]
I'm on the `filter-text-as-trigger` branch, but that should not matter.

**Comment by ayoub-benali** [comment]
Here I invoke the autocomplete each time only after the `T`
![Peek 2020-04-28 11-54](https://user-images.githubusercontent.com/1632384/80473885-094dd300-8947-11ea-860f-32005f3dab7a.gif)

> However, I get no completion results when the closing square bracket is missing

@olafurpg you can see that completion appear in each case

**Comment by rwols** [comment]
I believe it's due to https://github.com/sublimehq/sublime_text/issues/3299, sorry for the noise.
